File: cli/go/cmd/portfolio.go
package cmd

import (
	"encoding/json"
	"fmt"
	"os"

	"github.com/nexusconsult/nexus-cli/internal/client"
	"github.com/olekukonko/tablewriter"
	"github.com/spf13/cobra"
	"strings"
)

var portfolioCmd = &cobra.Command{
	Use:   "portfolio",
	Short: "Manage portfolio projects — add, remove, publish, organise",
	Long:  `Full CRUD management for the NexusConsult project portfolio.`,
}

// ── list ─────────────────────────────────────────────────────────────────────

var (
	portfolioListStatus   string
	portfolioListFeatured bool
	portfolioListFormat   string
)

var portfolioListCmd = &cobra.Command{
	Use:   "list",
	Short: "List all portfolio projects",
	Long: `Display all portfolio projects in a table or JSON.

Examples:
  nexus portfolio list
  nexus portfolio list --status active
  nexus portfolio list --featured
  nexus portfolio list --format json`,
	RunE: func(cmd *cobra.Command, args []string) error {
		var projects []map[string]interface{}
		if err := client.Get("/api/projects", &projects); err != nil {
			return err
		}
		// Filter
		filtered := make([]map[string]interface{}, 0)
		for _, p := range projects {
			if portfolioListStatus != "" && p["status"] != portfolioListStatus {
				continue
			}
			if portfolioListFeatured {
				if f, ok := p["featured"].(bool); !ok || !f {
					continue
				}
			}
			filtered = append(filtered, p)
		}

		if portfolioListFormat == "json" {
			out, _ := json.MarshalIndent(filtered, "", "  ")
			fmt.Println(string(out))
			return nil
		}

		t := tablewriter.NewWriter(os.Stdout)
		t.SetHeader([]string{"ID", "Name", "Type", "Status", "Published", "Featured", "Tags"})
		t.SetBorder(true)
		for _, p := range filtered {
			id   := fmt.Sprintf("%v", p["id"])
			if len(id) > 8 { id = id[:8] + "…" }
			tags := ""
			if tl, ok := p["tags"].([]interface{}); ok {
				ts := make([]string, 0, len(tl))
				for _, t := range tl { ts = append(ts, fmt.Sprintf("%v", t)) }
				tags = strings.Join(ts, ", ")
			}
			pub  := "✗"
			if v, ok := p["published"].(bool); ok && v { pub = "✓" }
			feat := "—"
			if v, ok := p["featured"].(bool);  ok && v { feat = "★" }
			t.Append([]string{
				id,
				fmt.Sprintf("%v", p["name"]),
				fmt.Sprintf("%v", p["type"]),
				fmt.Sprintf("%v", p["status"]),
				pub, feat, tags,
			})
		}
		t.Render()
		fmt.Printf("\n%s projects shown\n", dim(fmt.Sprintf("%d", len(filtered))))
		return nil
	},
}

// ── add ──────────────────────────────────────────────────────────────────────

var (
	addName        string
	addDescription string
	addType        string
	addTags        []string
	addGithub      string
	addStatus      string
	addPublished   bool
	addFeatured    bool
	addFromFile    string
)

var portfolioAddCmd = &cobra.Command{
	Use:   "add",
	Short: "Add a new project to the portfolio",
	Long: `Create a new portfolio project (requires admin token).

Examples:
  nexus portfolio add --name "MyService" --description "A great service" --type microservice --tags docker --tags k8s
  nexus portfolio add --from-file project.json`,
	RunE: func(cmd *cobra.Command, args []string) error {
		var payload map[string]interface{}

		if addFromFile != "" {
			data, err := os.ReadFile(addFromFile)
			if err != nil {
				return fmt.Errorf("read file: %w", err)
			}
			if err := json.Unmarshal(data, &payload); err != nil {
				return fmt.Errorf("parse JSON: %w", err)
			}
		} else {
			payload = map[string]interface{}{
				"name":        addName,
				"description": addDescription,
				"type":        addType,
				"tags":        addTags,
				"status":      addStatus,
				"published":   addPublished,
				"featured":    addFeatured,
			}
			if addGithub != "" {
				payload["githubUrl"] = addGithub
			}
		}

		var resp map[string]interface{}
		if err := client.Post("/api/projects", payload, &resp); err != nil {
			return err
		}
		id := fmt.Sprintf("%v", resp["id"])
		if len(id) > 8 { id = id[:8] + "…" }
		fmt.Printf("%s Project created: %s (id: %s)\n", green("✓"), cyan(fmt.Sprintf("%v", resp["name"])), id)
		return nil
	},
}

// ── update ────────────────────────────────────────────────────────────────────

var portfolioUpdateCmd = &cobra.Command{
	Use:   "update <project-id>",
	Short: "Update an existing project",
	Long: `Update portfolio project fields (requires admin token).

Examples:
  nexus portfolio update abc123 --status archived
  nexus portfolio update abc123 --name "New Name"`,
	Args: cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		payload := map[string]interface{}{}
		if cmd.Flags().Changed("name")   { payload["name"]   = addName }
		if cmd.Flags().Changed("description") { payload["description"] = addDescription }
		if cmd.Flags().Changed("status") { payload["status"] = addStatus }
		if cmd.Flags().Changed("github") { payload["githubUrl"] = addGithub }
		if cmd.Flags().Changed("tags")   { payload["tags"]   = addTags }
		if len(payload) == 0 {
			return fmt.Errorf("no fields to update — use --name, --status, --tags, etc.")
		}
		var resp map[string]interface{}
		if err := client.Patch("/api/projects/"+args[0], payload, &resp); err != nil {
			return err
		}
		fmt.Printf("%s Updated: %v\n", green("✓"), resp["name"])
		return nil
	},
}

// ── remove ────────────────────────────────────────────────────────────────────

var removeYes bool

var portfolioRemoveCmd = &cobra.Command{
	Use:   "remove <project-id>",
	Short: "Permanently remove a project",
	Long: `Delete a portfolio project (requires admin token).

Examples:
  nexus portfolio remove abc123
  nexus portfolio remove abc123 --yes`,
	Args: cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		if !removeYes {
			fmt.Printf("Delete project %s? [y/N]: ", args[0])
			var c string
			fmt.Scan(&c)
			if strings.ToLower(c) != "y" {
				fmt.Println("Aborted.")
				return nil
			}
		}
		if err := client.Delete("/api/projects/" + args[0]); err != nil {
			return err
		}
		fmt.Printf("%s Project %s deleted.\n", green("✓"), args[0])
		return nil
	},
}

// ── publish / unpublish / feature ─────────────────────────────────────────────

var portfolioPublishCmd = &cobra.Command{
	Use:   "publish <project-id>",
	Short: "Set project published=true",
	Long:  "Example:\n  nexus portfolio publish abc123",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		return client.Patch("/api/projects/"+args[0], map[string]bool{"published": true}, nil)
	},
}

var portfolioUnpublishCmd = &cobra.Command{
	Use:   "unpublish <project-id>",
	Short: "Set project published=false",
	Long:  "Example:\n  nexus portfolio unpublish abc123",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		return client.Patch("/api/projects/"+args[0], map[string]bool{"published": false}, nil)
	},
}

var featureOff bool
var portfolioFeatureCmd = &cobra.Command{
	Use:   "feature <project-id>",
	Short: "Toggle featured flag on a project",
	Long:  "Examples:\n  nexus portfolio feature abc123\n  nexus portfolio feature abc123 --off",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		return client.Patch("/api/projects/"+args[0], map[string]bool{"featured": !featureOff}, nil)
	},
}

// ── export ────────────────────────────────────────────────────────────────────

var exportOutput string

var portfolioExportCmd = &cobra.Command{
	Use:   "export",
	Short: "Export all projects to JSON file",
	Long: `Export all portfolio projects to a JSON file.

Examples:
  nexus portfolio export
  nexus portfolio export --output my-projects.json`,
	RunE: func(cmd *cobra.Command, args []string) error {
		var projects interface{}
		if err := client.Get("/api/projects", &projects); err != nil {
			return err
		}
		out, _ := json.MarshalIndent(projects, "", "  ")
		if err := os.WriteFile(exportOutput, out, 0644); err != nil {
			return err
		}
		fmt.Printf("%s Exported to %s\n", green("✓"), exportOutput)
		return nil
	},
}

func init() {
	portfolioListCmd.Flags().StringVarP(&portfolioListStatus,   "status",   "s", "",      "Filter by status (active/draft/archived)")
	portfolioListCmd.Flags().BoolVar(&portfolioListFeatured,    "featured",       false,   "Show only featured projects")
	portfolioListCmd.Flags().StringVarP(&portfolioListFormat,   "format",   "f", "table",  "Output format: table or json")

	portfolioAddCmd.Flags().StringVarP(&addName,        "name",        "n", "",       "Project name (required)")
	portfolioAddCmd.Flags().StringVarP(&addDescription, "description", "d", "",       "Short description (required)")
	portfolioAddCmd.Flags().StringVarP(&addType,        "type",        "t", "",       "Project type (required)")
	portfolioAddCmd.Flags().StringArrayVar(&addTags,    "tags",              nil,     "Tags (repeatable)")
	portfolioAddCmd.Flags().StringVar(&addGithub,       "github",            "",      "GitHub URL")
	portfolioAddCmd.Flags().StringVar(&addStatus,       "status",            "active","Status: active/draft/archived")
	portfolioAddCmd.Flags().BoolVar(&addPublished,      "published",         true,    "Published (default true)")
	portfolioAddCmd.Flags().BoolVar(&addFeatured,       "featured",          false,   "Featured")
	portfolioAddCmd.Flags().StringVar(&addFromFile,     "from-file",         "",      "Load from JSON file")

	portfolioUpdateCmd.Flags().StringVarP(&addName,        "name",        "n", "", "New name")
	portfolioUpdateCmd.Flags().StringVarP(&addDescription, "description", "d", "", "New description")
	portfolioUpdateCmd.Flags().StringVar(&addStatus,       "status",            "", "New status")
	portfolioUpdateCmd.Flags().StringVar(&addGithub,       "github",            "", "New GitHub URL")
	portfolioUpdateCmd.Flags().StringArrayVar(&addTags,    "tags",              nil, "Replace tags")

	portfolioRemoveCmd.Flags().BoolVarP(&removeYes, "yes", "y", false, "Skip confirmation")
	portfolioFeatureCmd.Flags().BoolVar(&featureOff, "off", false, "Unfeature the project")
	portfolioExportCmd.Flags().StringVarP(&exportOutput, "output", "o", "portfolio-export.json", "Output file path")

	portfolioCmd.AddCommand(
		portfolioListCmd, portfolioAddCmd, portfolioUpdateCmd,
		portfolioRemoveCmd, portfolioPublishCmd, portfolioUnpublishCmd,
		portfolioFeatureCmd, portfolioExportCmd,
	)
}

File: cli/python/commands/portfolio.py
"""
nexus portfolio — Portfolio project management commands.

Commands:
  list      List all portfolio projects
  get       Get a project by ID or name
  add       Add (create) a new project to the portfolio
  update    Update an existing project's fields
  remove    Remove a project from the portfolio
  publish   Set a project to published=true
  unpublish Set a project to published=false
  feature   Toggle a project's featured flag
  import    Bulk import projects from a JSON file
  export    Export all projects to a JSON file
"""

import json
from pathlib import Path
from typing import Optional

import click
from rich.console import Console
from rich.table import Table
from rich.panel import Panel
from rich import box

from ..lib import client

console = Console()

STATUS_COLOR = {"active": "green", "draft": "yellow", "archived": "dim"}


def _status_badge(s):
    color = STATUS_COLOR.get(s, "white")
    return f"[{color}]{s}[/]"


def _print_project(p: dict):
    t = Table(show_header=False, box=box.ROUNDED, border_style="cyan")
    t.add_column("Field", style="cyan",  min_width=18)
    t.add_column("Value", style="white")
    for k, v in p.items():
        if k == "tags" and isinstance(v, list):
            v = ", ".join(v) or "—"
        t.add_row(k, str(v))
    console.print(t)


@click.group()
def portfolio():
    """Manage portfolio projects — add, remove, publish, and organise."""
    pass


# ── list ─────────────────────────────────────────────────────────────────────

@portfolio.command("list")
@click.option("--status",   "-s", type=click.Choice(["active","draft","archived"]), default=None)
@click.option("--featured", is_flag=True, help="Show only featured projects.")
@click.option("--format",   "-f", "fmt", type=click.Choice(["table","json"]), default="table")
@click.help_option("--help", "-h")
def list_projects(status, featured, fmt):
    """
    List all portfolio projects.

    \b
    Examples:
      nexus portfolio list
      nexus portfolio list --status active
      nexus portfolio list --featured
      nexus portfolio list --format json
    """
    try:
        projects = client.get("/api/projects")
    except client.APIError as e:
        console.print(f"[red]Error:[/] {e.detail}"); return

    if status:
        projects = [p for p in projects if p.get("status") == status]
    if featured:
        projects = [p for p in projects if p.get("featured")]

    if fmt == "json":
        console.print_json(json.dumps(projects, indent=2, default=str))
        return

    t = Table(title=f"Portfolio Projects ({len(projects)})", box=box.ROUNDED, border_style="blue")
    t.add_column("ID",          style="dim",     min_width=8)
    t.add_column("Name",        style="bold cyan", min_width=22)
    t.add_column("Type",        style="white",    min_width=14)
    t.add_column("Status",      min_width=10)
    t.add_column("Published",   justify="center", min_width=10)
    t.add_column("Featured",    justify="center", min_width=10)
    t.add_column("Tags",        style="dim")

    for p in projects:
        pub_icon = "[green]✓[/]" if p.get("published") else "[red]✗[/]"
        feat_icon = "[yellow]★[/]" if p.get("featured") else "—"
        t.add_row(
            p.get("id","")[:8] + "…",
            p.get("name",""),
            p.get("type",""),
            _status_badge(p.get("status","active")),
            pub_icon, feat_icon,
            ", ".join(p.get("tags",[])[:3]),
        )
    console.print(t)


# ── get ──────────────────────────────────────────────────────────────────────

@portfolio.command()
@click.argument("project_id")
@click.help_option("--help", "-h")
def get(project_id):
    """
    Show full details of a project by its ID.

    \b
    Examples:
      nexus portfolio get abc123
    """
    try:
        projects = client.get("/api/projects")
        match    = next((p for p in projects if p["id"] == project_id or project_id in p["id"]), None)
        if not match:
            console.print(f"[red]Project not found:[/] {project_id}"); return
        _print_project(match)
    except client.APIError as e:
        console.print(f"[red]Error:[/] {e.detail}")


# ── add ──────────────────────────────────────────────────────────────────────

@portfolio.command()
@click.option("--name",        "-n", required=True,  help="Project name.")
@click.option("--description", "-d", required=True,  help="Short description.")
@click.option("--type",        "-t", required=True,  help="Project type (e.g. microservice, ML, infra).")
@click.option("--tags",        multiple=True,         help="Tags (repeatable): --tags docker --tags k8s")
@click.option("--github",                            help="GitHub repository URL.")
@click.option("--status",      default="active",     type=click.Choice(["active","draft","archived"]))
@click.option("--published",   is_flag=True, default=True)
@click.option("--featured",    is_flag=True, default=False)
@click.option("--long-desc",   help="Long description (markdown supported).")
@click.option("--run-cmd",     help="Example run command.")
@click.option("--test-cmd",    help="Example test command.")
@click.option("--from-file",   type=click.Path(exists=True), help="Load project fields from JSON file.")
@click.help_option("--help", "-h")
def add(name, description, type, tags, github, status, published, featured,
        long_desc, run_cmd, test_cmd, from_file):
    """
    Add a new project to the portfolio (requires admin token).

    \b
    Examples:
      nexus portfolio add --name "MyService" --description "A great service" --type microservice --tags docker --tags k8s
      nexus portfolio add --from-file project.json
    """
    if from_file:
        payload = json.loads(Path(from_file).read_text())
    else:
        payload = {
            "name": name, "description": description, "type": type,
            "tags": list(tags), "status": status, "published": published,
            "featured": featured,
        }
        if github:   payload["githubUrl"]      = github
        if long_desc: payload["longDescription"] = long_desc
        if run_cmd:  payload["runCommand"]     = run_cmd
        if test_cmd: payload["testCommand"]    = test_cmd

    try:
        data = client.post("/api/projects", payload)
        console.print(f"[green]✓ Project created:[/] [cyan]{data['name']}[/] (id: {data['id'][:8]}…)")
    except client.APIError as e:
        console.print(f"[red]Error:[/] {e.detail}"); raise SystemExit(1)


# ── update ────────────────────────────────────────────────────────────────────

@portfolio.command()
@click.argument("project_id")
@click.option("--name",        help="New name.")
@click.option("--description", help="New description.")
@click.option("--status",      type=click.Choice(["active","draft","archived"]))
@click.option("--github",      help="New GitHub URL.")
@click.option("--tags",        multiple=True, help="Replace tags.")
@click.help_option("--help", "-h")
def update(project_id, name, description, status, github, tags):
    """
    Update fields of an existing project (requires admin token).

    \b
    Examples:
      nexus portfolio update abc123 --status archived
      nexus portfolio update abc123 --name "New Name" --tags api --tags rest
    """
    payload: dict = {}
    if name:        payload["name"] = name
    if description: payload["description"] = description
    if status:      payload["status"] = status
    if github:      payload["githubUrl"] = github
    if tags:        payload["tags"] = list(tags)
    if not payload:
        console.print("[yellow]No fields to update.[/]"); return
    try:
        data = client.patch(f"/api/projects/{project_id}", payload)
        console.print(f"[green]✓ Updated:[/] {data.get('name','—')}")
    except client.APIError as e:
        console.print(f"[red]Error:[/] {e.detail}"); raise SystemExit(1)


# ── remove ────────────────────────────────────────────────────────────────────

@portfolio.command()
@click.argument("project_id")
@click.option("--yes", "-y", is_flag=True, help="Skip confirmation.")
@click.help_option("--help", "-h")
def remove(project_id, yes):
    """
    Permanently remove a project from the portfolio (requires admin token).

    \b
    Examples:
      nexus portfolio remove abc123
      nexus portfolio remove abc123 --yes
    """
    if not yes:
        click.confirm(f"Delete project {project_id}?", abort=True)
    try:
        client.delete(f"/api/projects/{project_id}")
        console.print(f"[green]✓ Project {project_id} deleted.[/]")
    except client.APIError as e:
        console.print(f"[red]Error:[/] {e.detail}"); raise SystemExit(1)


# ── publish / unpublish ───────────────────────────────────────────────────────

@portfolio.command()
@click.argument("project_id")
@click.help_option("--help", "-h")
def publish(project_id):
    """Set project published=true (requires admin).\n\nExample: nexus portfolio publish abc123"""
    try:
        client.patch(f"/api/projects/{project_id}", {"published": True})
        console.print(f"[green]✓ Project {project_id} is now published.[/]")
    except client.APIError as e:
        console.print(f"[red]Error:[/] {e.detail}")


@portfolio.command()
@click.argument("project_id")
@click.help_option("--help", "-h")
def unpublish(project_id):
    """Set project published=false (requires admin).\n\nExample: nexus portfolio unpublish abc123"""
    try:
        client.patch(f"/api/projects/{project_id}", {"published": False})
        console.print(f"[green]✓ Project {project_id} unpublished.[/]")
    except client.APIError as e:
        console.print(f"[red]Error:[/] {e.detail}")


@portfolio.command()
@click.argument("project_id")
@click.option("--off", is_flag=True, help="Unfeature instead of feature.")
@click.help_option("--help", "-h")
def feature(project_id, off):
    """Toggle featured flag on a project.\n\nExample: nexus portfolio feature abc123"""
    try:
        client.patch(f"/api/projects/{project_id}", {"featured": not off})
        console.print(f"[green]✓ Project {project_id} {'un' if off else ''}featured.[/]")
    except client.APIError as e:
        console.print(f"[red]Error:[/] {e.detail}")


# ── import / export ───────────────────────────────────────────────────────────

@portfolio.command("import")
@click.argument("json_file", type=click.Path(exists=True))
@click.option("--dry-run", is_flag=True, help="Validate without creating.")
@click.help_option("--help", "-h")
def import_projects(json_file, dry_run):
    """
    Bulk import projects from a JSON file (list of project objects).

    \b
    Examples:
      nexus portfolio import projects.json
      nexus portfolio import projects.json --dry-run
    """
    projects = json.loads(Path(json_file).read_text())
    if not isinstance(projects, list):
        projects = [projects]
    console.print(f"[cyan]Importing {len(projects)} projects{' (dry run)' if dry_run else ''}…[/]")
    for p in projects:
        if dry_run:
            console.print(f"  [dim]would create:[/] {p.get('name','?')}")
            continue
        try:
            data = client.post("/api/projects", p)
            console.print(f"  [green]✓[/] {data['name']}")
        except client.APIError as e:
            console.print(f"  [red]✗[/] {p.get('name','?')}: {e.detail}")


@portfolio.command("export")
@click.option("--output", "-o", default="portfolio-export.json", show_default=True)
@click.help_option("--help", "-h")
def export_projects(output):
    """
    Export all portfolio projects to a JSON file.

    \b
    Examples:
      nexus portfolio export
      nexus portfolio export --output my-projects.json
    """
    try:
        projects = client.get("/api/projects")
        Path(output).write_text(json.dumps(projects, indent=2, default=str))
        console.print(f"[green]✓ Exported {len(projects)} projects to {output}[/]")
    except client.APIError as e:
        console.print(f"[red]Error:[/] {e.detail}")
